src/flint/types/fq_default_poly.pyx

```diff
@@ -0,0 +1,1643 @@
+
+        cdef fq_default_poly res
+        res = self.ctx.new_ctype_poly()
+        fq_default_poly_sub_series(
+            res.val, self.val, other_c.val, n, res.ctx.field.val
+        )
+        return res
+
+    def mul_low(self, other, slong n):
+        r"""
+        Returns the lowest ``n`` coefficients of the multiplication of ``self`` with ``other``
+
+        Equivalent to computing `f(x) \cdot g(x) \mod x^n`
+
+            >>> R = fq_default_poly_ctx(163)
+            >>> f = R([2,3,5,7,11])
+            >>> g = R([1,2,4,8,16])
+            >>> f.mul_low(g, 5)
+            101*x^4 + 45*x^3 + 19*x^2 + 7*x + 2
+            >>> f.mul_low(g, 3)
+            19*x^2 + 7*x + 2
+            >>> f.mul_low(g, 1)
+            2
+        """
+        # Only allow multiplication with other fq_default_poly
+        if not typecheck(other, fq_default_poly):
+            raise TypeError("other polynomial must be of type fq_default_poly")
+
+        # Ensure the contexts match
+        other_c = <fq_default_poly>other
+        if self.ctx != other_c.ctx:
+            raise ValueError("other polynomial's context does not match")
+
+        cdef fq_default_poly res
+        res = self.ctx.new_ctype_poly()
+        fq_default_poly_mullow(
+            res.val, self.val, other_c.val, n, res.ctx.field.val
+        )
+        return res
+
+    def pow_trunc(self, slong e, slong n):
+        """
+        Returns ``self`` raised to the power ``e`` modulo `x^n`:
+        :math:`f^e \mod x^n`/
+
+        Note: For exponents larger that 2^31 (which do not fit inside a ulong) use the
+        method :method:`~.pow_mod` with the explicit modulus `x^n`.
+
+            >>> R = fq_default_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = 30*x**6 + 104*x**5 + 76*x**4 + 33*x**3 + 70*x**2 + 44*x + 65
+            >>> f.pow_trunc(2**20, 30) == pow(f, 2**20, x**30)
+            True
+            >>> f.pow_trunc(2**20, 5)
+            132*x^4 + 113*x^3 + 36*x^2 + 48*x + 6
+        """
+        if e < 0:
+            raise ValueError("Exponent must be non-negative")
+
+        cdef fq_default_poly res
+        res = self.ctx.new_ctype_poly()
+        fq_default_poly_pow_trunc(res.val, self.val, e, n, res.ctx.field.val)
+        return res
+
+    def sqrt_trunc(self, slong n):
+        """
+        Returns the squareroot of ``self`` modulo `x^n`.
+
+            >>> R = fq_default_poly_ctx(163, 3)
+            >>> x = R.gen()
+            >>> z = R.base_field().gen()
+            >>> f = (37*z + 54)*x**3 + (8*z + 94)*x**2 + (52*z + 142)*x + 1
+            >>> h = f.sqrt_trunc(4)
+            >>> h
+            (7*z^2 + 17*z + 148)*x^3 + (151*z^2 + 114*z + 53)*x^2 + (26*z + 71)*x + 1
+            >>> h.mul_low(h, 4) == f
+            True
+        """
+        cdef fq_default_poly res
+
+        # FLINT assumes the constant term is one
+        c = self.constant_coefficient()
+        if not c.is_one():
+            self = self / c
```

Comment:
Will this not change the value of the sqrt? I would have thought that we could only do this if we can multiply by `sqrt(c)` after...

Reply:
Oh you're probably right, it might be easier to simply require that the constant coefficient is set to 1? If the constant term has no square root I suppose this is ill defined.

Reply:
If `p(x)` has constant term `c` then `p(x)^2` has constant term `c^2`. So I think that if the sqrt exists then the sqrt of the constant term does as well.

Reply:
It's different with the series calculation though? As in I think this function always "succeeds" as long as the constant term is 1, but the sqrt will just be an approximation?

Reply:
We also have:
```python
In [5]: x = fmpq_series([0, 1])

In [6]: x.sqrt()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line 1
----> 1 x.sqrt()

File fmpq_series.pyx:380, in flint.types.fmpq_series.fmpq_series.sqrt()

ValueError: sqrt() of power series: constant term != 1
```
I suppose that if the ground domain is a field then we can always compute a series for `(1 + a*x + b*x^2 + ...)^(1/2)` through binomal expansion. More generally if the constant term `c` is square then we can divide by it to get a series like `sqrt(p/c)*sqrt(c)`.

Basically the first term of the series needs to be `sqrt(c)` so the series only exists if `sqrt(c)` does.

Reply:
Okay I agree with this. What I'll do (probably tomorrow) is edit these two functions to check whether the constant term is a square instead of 1. This is cleaner and uses that the base of the polynomial is a field :)